src/core/algorithms/gfd/gfd_miner.cpp

```diff
@@ -0,0 +1,667 @@
+#include <boost/graph/isomorphism.hpp>
+#include <boost/graph/vf2_sub_graph_iso.hpp>
+#include <easylogging++.h>
+
+#include "config/option_using.h"
+
+namespace {
+
+using namespace algos;
+
+using Embedding = std::map<vertex_t, vertex_t>;
+using Embeddings = std::vector<Embedding>;
+
+using SingleRule = std::pair<std::vector<Literal>, Literal>;
+using Rule = std::pair<std::vector<Literal>, std::vector<Literal>>;
+using Rules = std::vector<Rule>;
+
+using Info = std::map<std::string, std::set<std::string>>;
+
+void NextSubset(std::vector<std::size_t>& indices, std::size_t const border) {
+    if (indices.at(0) == border - indices.size()) {
+        indices = {};
+        return;
+    }
+    std::size_t index = 0;
+    for (int i = static_cast<int>(indices.size()) - 1; i >= 0; --i) {
+        if (indices.at(i) != border - indices.size() + static_cast<size_t>(i)) {
+            index = i;
+            break;
+        }
+    }
+    indices.at(index)++;
+    for (std::size_t i = index + 1; i < indices.size(); ++i) {
+        indices.at(i) = indices.at(index) + i - index;
+    }
+}
+
+template <typename T>
+std::vector<std::vector<T>> GetSubsets(std::vector<T> const& elements, std::size_t const n) {
+    if (elements.size() < n) {
+        return {};
+    }
+    std::vector<std::vector<T>> result = {};
+    std::vector<std::size_t> indices = std::vector<std::size_t>(n);
+    std::iota(std::begin(indices), std::end(indices), 0);
+    while (!indices.empty()) {
+        std::vector<T> element = {};
+        for (std::size_t index : indices) {
+            element.push_back(elements.at(index));
+        }
+        result.push_back(element);
+        NextSubset(indices, elements.size());
+    }
+    return result;
+}
+
+bool Satisfied(std::vector<Literal> const& literals, graph_t const& graph, graph_t const& pattern,
+               Embedding const& embedding) {
+    for (Literal const& l : literals) {
+        Token fst_token = l.first;
+        Token snd_token = l.second;
+        std::string fst, snd;
+        if (fst_token.first == -1) {
+            fst = fst_token.second;
+        } else {
+            vertex_t u = boost::vertex(fst_token.first, pattern);
+            vertex_t v = embedding.at(u);
+            std::map<std::string, std::string> attrs = graph[v].attributes;
+            if (attrs.find(fst_token.second) == attrs.end()) {
+                return false;
+            }
+            fst = attrs.at(fst_token.second);
+        }
+        if (snd_token.first == -1) {
+            snd = snd_token.second;
+        } else {
+            vertex_t u = boost::vertex(snd_token.first, pattern);
+            vertex_t v = embedding.at(u);
+            std::map<std::string, std::string> attrs = graph[v].attributes;
+            if (attrs.find(snd_token.second) == attrs.end()) {
+                return false;
+            }
+            snd = attrs.at(snd_token.second);
+        }
+        if (fst != snd) {
+            return false;
+        }
```

Comment:
I'd create a function
```suggestion
bool checkToken(const Token& token, std::string& tok, graph_t const& graph, graph_t const& pattern,
               Embedding const& embedding) {
        if (fst_token.first == -1) {
            tok = token.second;
        } else {
            vertex_t u = boost::vertex(token.first, pattern);
            vertex_t v = embedding.at(u);
            std::map<std::string, std::string> attrs = graph[v].attributes;
            if (attrs.find(token.second) == attrs.end()) {
                return false;
            }
            tok = attrs.at(token.second);
        }
        return true;
}

bool Satisfied(std::vector<Literal> const& literals, graph_t const& graph, graph_t const& pattern,
               Embedding const& embedding) {
    for (Literal const& l : literals) {
        Token fst_token = l.first;
        Token snd_token = l.second;
        std::string fst, snd;
        bool first_res = checkToken(fst_token, fst, graph, pattern, embedding);
        bool second_res = checkToken(snd_token, snd, graph, pattern, embedding);
        if (!first_res or !second_res or fst != snd)
            return false;
}
```